crates/wast/src/resolve/names.rs

```diff
@@ -1742,87 +1828,62 @@ impl<'a, 'b> ExprResolver<'a, 'b> {
-                // results.
-                //
-                // In MVP wasm blocks can have 0 params and 0-1 results. Now
-                // there's also multi-value. We want to prefer MVP wasm wherever
-                // possible (for backcompat) so we want to list this block as
-                // being an "MVP" block if we can. The encoder only has
-                // `BlockType` to work with, so it'll be looking at `params` and
-                // `results` to figure out what to encode. If `params` and
-                // `results` fit within MVP, then it uses MVP encoding
-                //
-                // To put all that together, here we handle:
-                //
-                // * If the `index` was specified, resolve it and use it as the
-                //   source of truth. If this turns out to be an MVP type,
-                //   record it as such.
-                // * Otherwise use `params` and `results` as the source of
-                //   truth. *If* this were a non-MVP compatible block `index`
-                //   would be filled by by `tyexpand.rs`.
-                //
-                // tl;dr; we handle the `index` here if it's set and then fill
-                // out `params` and `results` if we can, otherwise no work
-                // happens.
-                if bt.ty.index.is_some() {
-                    let (ty, _) = self.module.resolve_type_use(&mut bt.ty)?;
-                    let n = match ty {
-                        Index::Num(n, _) => *n,
-                        Index::Id(_) => panic!("expected `Num`"),
-                    };
-                    let ty = match self
-                        .module
-                        .types
-                        .items
-                        .get(n as usize)
-                        .and_then(|s| s.item())
-                    {
-                        Some(TypeInfo::Func(ty)) => ty,
-                        _ => return Ok(()),
-                    };
-                    if ty.0.len() == 0 && ty.1.len() <= 1 {
-                        let mut inline = FunctionType::default();
-                        inline.results = ty.1.clone();
-                        bt.ty.inline = Some(inline);
-                        bt.ty.index = None;
-                    }
+            Let(t) => {
+                // Resolve (ref T) in locals
+                for local in &mut t.locals {
+                    self.module.resolve_valtype(&mut local.ty)?;
                 }
 
-                // If the inline annotation persists to this point then resolve
-                // all of its inline value types.
-                if let Some(inline) = &mut bt.ty.inline {
-                    inline.resolve(self.module)?;
+                // Register all locals defined in this let
+                let mut scope = Namespace::default();
+                for local in &t.locals {
+                    scope.register(local.id, "local")?;
```

Comment:
The semantics of this implementation I think means that you can redefine a local name multiple times, it'll just shadow within a scope. Do you know if that's the intention of the `let` proposal? Or is it intended that shadowing is not allowed and redefining a name should be an error?

Reply:
Yes, that appears to be the intent of the proposal. I originally had it the other way around, and it failed a couple tests in `let.wast` which checked that shadowing was the behavior.